### PR TITLE
Memory resource design

### PR DESCRIPTION
**Pull request by jrhemstad**
Based on [`std::pmr::memory_resource`](https://en.cppreference.com/w/cpp/memory/memory_resource) and [`thrust::mr::memory_resource`](https://github.com/thrust/thrust/blob/master/thrust/mr/memory_resource.h).

Key differences:

- Doesn't have an `alignment` argument. It will be assumed that all implementations of `device_memory_resource` will provide at least 256B alignment.
- Adds a `stream` argument for optional allocation/deallocation within a CUDA stream (e.g., CNMEM supports streams in alloc/free)
- Adds a `supports_streams` pure virtual function to query if the resource supports streams

In short, the `device_memory_resource` is a base class that has two public functions: `allocate` and `deallocate`. These functions call corresponding private, pure virtual functions `do_allocate` and `do_deallocate`. 

To define a custom memory resource, all that need be done is define a new class, derive from `device_memory_resource`, and implement `do_allocate` and `do_deallocate` to perform whatever custom behavior is desired. To see an example, look at how I've implemented `cuda_memory_resource`, which just uses `cudaMalloc` and `cudaFree` for alloc/free.

Example usage:

```
class my_custom_resource : public device_memory_resource { 

private:

void * do_allocate(size_t ptr, cudaStream_t stream) override { ... };

void do_deallocate(void * ptr, cudaStream_t stream) override {...};
};

device_memory_resource * base_resource_ptr = new my_custom_resource();

// `output` will be allocated using the custom memory resource
column output = cudf::some_function(..., base_resource_ptr);
```

By using runtime polymorphism via virtual functions in `device_memory_resource`, this design allows us to keep libcudf APIs template-free while still providing the option to the user to customize memory allocation behavior.

TODOs:

- [x] Add default memory resource functionality

### CODE REVIEW

File: device_memory_resource.cuh [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * If supported, this operation may optionally be executed on a stream.
+   * Otherwise, the stream is ignored and the null stream is used.
+   *
+   * @param bytes The size of the allocation
+   * @param stream Stream on which to perform allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  virtual void* do_allocate(std::size_t bytes, cudaStream_t stream) = 0;
```
Comment (revans2) [inline device_memory_resource.cuh]: How are errors returned?  throwing std::bad_alloc?  returning NULL?  We probably want some clear documentation beyond the example that ignores the detailed errors and returns NULL.

**Comment by jlowe** [comment]
Thanks for the writeup!  This looks workable to me.

I assume the allocator is tracked by each `device_buffer` rather than the column itself?  I'm thinking of [the nans-as-nulls example](https://github.com/jrhemstad/cudf_column/pull/1#discussion_r288749290) where code might be handed a data column and compute its own validity bitmask not realizing the data column wasn't using the default allocator.

**Comment by jrhemstad** [comment]
> I assume the allocator is tracked by each `device_buffer` rather than the column itself? I'm thinking of [the nans-as-nulls example](https://github.com/jrhemstad/cudf_column/pull/1#discussion_r288749290) where code might be handed a data column and compute its own validity bitmask not realizing the data column wasn't using the default allocator.

Correct, each `device_buffer` object will have an associated `device_memory_resource*`. I don't see a reason yet for the column itself to track it's own allocator separate from what is associated with it's constituent `device_buffer`s

**Review by revans2 — COMMENTED** [review]
From a high level it looks fine for what it was designed to do. I like that there is the option to wrap other memory resource instances so we could have some kind of generic debugging separate from RMM to be able to help find memory leaks. But I would really like to see how it fits in with some example CUDF APIs.

The main use case we have for this is trying to reduce/coalese small data transfers specifically when exchanging data between different machines through the host.

Typically this will happen like the following.
1) Partition the data on the GPU
2) **Transfer the data to the host**
3) Write the data out to some other java API that will do the transfer (each partition will go to a different machine)
4) On a separate system read the data in and store it in host memory.
5) **Transfer the data to the device**
6) Setup columns to point to the data we transferred.  

For step 5 we can do it in a single transfer per partition being read in.  We can do this by allocating all of the memory we would need for that partition up front, and then only deallocating it when all of the columns are done being used.  I can totally see how to make that work with this current set of APIs, we would only have to do some small bookkeeping to know when we released everything and can finally release the overall memory.

For step 2 it would be ideal to be able to make it a single transfer, but it would require that we have the output of partition be in a single contiguous block.  I am a bit fuzzy on how to make it work if there is temporary data allocated using the same device_memory_resource during the partition phase.  We can allocate a block large enough to hold all of the output data up front and hand out sections of it when asked, but I don't know how to keep the temporary data in a separate arena.

It would be great to see an example partition API that would give me something more concrete to work on/think about.

File: device_memory_resource.cuh (hunk 1 above)
Comment (jrhemstad) [reply device_memory_resource.cuh]: Errors will be handled via thrown exceptions. I omitted that in this first draft for simplicity, but will be adding them soon.

**Comment by jrhemstad** [comment]
> For step 2 it would be ideal to be able to make it a single transfer, but it would require that we have the output of partition be in a single contiguous block. I am a bit fuzzy on how to make it work if there is temporary data allocated using the same device_memory_resource during the partition phase. We can allocate a block large enough to hold all of the output data up front and hand out sections of it when asked, but I don't know how to keep the temporary data in a separate arena.

@revans2 Hmmm, this is good feedback. Perhaps libcudf APIs need separate `device_memory_resource`s for output allocations and temporary allocations. 

I'm not in love with that idea because it does increase complexity all around, so I want to challenge it a bit.

Effectively you're wanting to roll your own "pool" allocation by allocating a big chunk of memory up-front and then handing out contiguous pieces of it. That's a sensible performance optimization. However, let's suppose that RMM pool allocation is fast enough that doing your own "pool" doesn't move the needle much. Would you be content with this?

Switching sides and arguing _for_ separate output/temp allocators. I could imagine a scenario in which perhaps you want your output allocated with UVM, but it would be unnecessary overhead to have temporary allocations use UVM. Same could be true of pinned host memory vs. unpinned. 

I can certainly imagine a number of scenarios where you'd really want to control output allocations independently of temporary allocations. However, I just want to make sure that those scenarios are _real_ enough to warrant the added complexity of having separate memory resources. I can already envision bugs that result from libcudf functions using the wrong memory resource for a temporary allocation.

**Comment by jlowe** [comment]
I didn't think partitioning was an issue because I assumed (possibly incorrectly) that the custom allocator was _only_ for output columns visible to the caller.  If temporary memory was needed by a cudf operation, I assumed the specified allocator was _not_ going to be used for those temporary allocations.  Otherwise yeah, it's going to be too hard for the custom allocator to do any kind of special placement logic on the output vector since it won't know whether the allocation is for the final output or a temporary that will still be alive at the time the output needs to be allocated.

**Comment by jrhemstad** [comment]
> I didn't think partitioning was an issue because I assumed (possibly incorrectly) that the custom allocator was _only_ for output columns visible to the caller. If temporary memory was needed by a cudf operation, I assumed the specified allocator was _not_ going to be used for those temporary allocations. Otherwise yeah, it's going to be too hard for the custom allocator to do any kind of special placement logic on the output vector since it won't know whether the allocation is for the final output or a temporary that will still be alive at the time the output needs to be allocated.

What would you want to see used for temporary memory allocation then? 

To elaborate a bit further on the design I have in mind:

There will be a "default" global memory resource that you can access via `get/set_default_memory_resource`. Every memory resource function parameter will default to ` = get_default_memory_resource()`. This way, if you just want to configure _all_ allocation to use a specific custom resource, it is easy to do so via `set_default_memory_resource`. For example, to enable RMM pool mode, it would just be a call to `set_default_memory_resource(&pool_device_memory_resource)`.

Returning to temporary memory allocations...

Would you want libcudf APIs to accept separate memory resources for temp vs output allocations? There could be a _global_ temporary default memory resource independent from the default memory resource I described above.  

Would it be a happy compromise to allow configuring the _global_ temporary resource, but only accept memory resources for _output_ allocations on a per-API basis?

**Comment by jlowe** [comment]
> This way, if you just want to configure all allocation to use a specific custom resource, it is easy to do so via `set_default_memory_resource`

Yes, I expect there would be some way to set a global, default allocator across all threads that is used for all temporary allocations and final outputs for operations that did not have a custom allocator specified.  I do not have expectations that users can control temporary allocations for a cudf operation separate from the global allocator.  There would only be control of an operation's final output allocations on a per-operation basis.

I don't see a use-case for needing control over temporary allocations separate from the default allocator on a per-operation basis.  If someone needs the ability to control that in the future then either the API could be extended to take per-operation temporary allocator arguments, or maybe there could be a way to set a thread-local default allocator override.  But I don' think we need to get into that yet, if ever.

**Comment by revans2** [comment]
So, first of all, I am not really concerned about the allocation speed of the pool.  I assume that RMM is fast enough so I am not worried about that.  What I am interested in is the placement of the data. In some cases, I would like the output to be in a contiguous block so I can transfer it much more efficiently.

Facebook shared some numbers in a talk at the last spark/ai summit that showed for them each partition was typically a few 100 KB in size.  That data is compressed and in a different format, but in general, it is a small amount of data, and if we have to subdivide it further by the number of columns and validity vectors we are talking about tens of transfers each in the tens of KB in size.  We really want to avoid this because of the overhead in starting a DMA transfer.

We can fix some of this with education, Spark has a default of 200 partitions which is exacerbating the problem, but we also want to do a decent job even in a bad situation.

**Comment by revans2** [comment]
I agree with @jlowe I don't see a reason to control temporary allocations separate from the final outputs, and for me, I don't see much value in controlling the output locations except in a few special situations, when we know that we are going to transfer data.

**Comment by jrhemstad** [comment]
@jlowe @revans2 okay, so to be clear, you'd be happy with the following:

1. There is a _single_ global `device_memory_resource` that by default will be used for _both_ temporary and outpout memory allocation.
   - This global resource can be accessed and modified via appropriate `get/set` functions
2. On a per-function basis, _output_ allocation can optionally be customized via a `device_memory_resource*` parameter (that defaults to the global).
   - There will be no per-function control over temporary memory allocation

**Comment by revans2** [comment]
That would be fine for me, but you might want to check with others that wanted to control allocation to be sure that their use case fits with the one that we have.

**Comment by jrhemstad** [comment]
Closing. See https://github.com/rapidsai/rmm/pull/96 for final design and tests.

**Pull request closed without merging**